jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
+        private final AtomicReference<LastWill> cancelled;
+        private final AtomicLong demand;
+        private Content.Source content;
+        private Flow.Subscriber<? super Content.Chunk> subscriber;
 
-                if (chunk == null)
+        public ActiveSubscription(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+        {
+            this.cancelled = new AtomicReference<>(null);
+            this.demand = new AtomicLong(0);
+            this.content = content;
+            this.subscriber = subscriber;
+        }
+
+        // As per rule 3.3, Subscription MUST place an upper bound on possible synchronous
+        // recursion between Publisher and Subscriber
+        //
+        // As per rule 1.3, onSubscribe, onNext, onError and onComplete signaled to a
+        // Subscriber MUST be signaled serially.
+        //
+        // IteratingCallback guarantee that process() method will be executed by one thread only.
+        // The process() method can be only initiated from request() or cancel() demands methods.
+        @Override
+        protected Action process()
+        {
+            LastWill cancelled = this.cancelled.get();
+            if (cancelled != null)
+            {
+                Throwable reason = cancelled.reason;
+                FinalSignal finalSignal = cancelled.finalSignal;
+                // As per rule 3.13, Subscription.cancel() MUST request the Publisher to eventually
+                // drop any references to the corresponding subscriber.
+                this.demand.set(NO_MORE_DEMAND);
+                // TODO: HttpChannelState does not satisfy the contract of Content.Source "If read() has returned a last chunk, this is a no operation."
+                if (finalSignal != FinalSignal.COMPLETE)
+                    this.content.fail(reason);
+                this.content = null;
+                try
                 {
-                    try (AutoLock ignored = lock.lock())
-                    {
-                        // Restore the demand decremented above.
-                        ++demand;
-                        stalled = true;
-                    }
-                    content.demand(this::process);
-                    return;
+                    if (finalSignal == FinalSignal.COMPLETE)
+                        this.subscriber.onComplete();
+                    if (finalSignal == FinalSignal.ERROR)
+                        this.subscriber.onError(reason);
+                    // do nothing on FinalSignal.SUPPRESS
                 }
-
-                if (Content.Chunk.isFailure(chunk))
+                catch (Throwable err)
                 {
-                    terminate();
-                    if (!chunk.isLast())
-                        content.fail(chunk.getFailure());
-                    subscriber.onError(chunk.getFailure());
-                    return;
+                    LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
```

Comment:
Why are you catching Throwable here?

Reply:
@lorban ActiveSubscription.process() method triggered from Subscription.cancel/request methods and may be executed synchronously
- 3.10, 3.11 While the Subscription is not cancelled, Subscription.request(long n) MAY synchronously call onNext on this
- 3.15 and 3.16 Calling Subscription.cancel/request MUST return normal (that's the reason why we cannot throw here)
- 2.13, ...the caller MUST raise this error condition in a fashion that is adequate for the runtime environment. «Raise this error condition in a fashion that is adequate for the runtime environment» could mean logging the error—or otherwise make someone or something aware of the situation—as the error cannot be signalled to the faulty Subscriber.